server/controllers/finance/reports/balance/index.js

```diff
@@ -0,0 +1,227 @@
+    var obj = account[id] = {}, sold = getSold(row);
+    obj.label = row.label;
+    obj.beginDebit = sold.debit;
+    obj.beginCredit = sold.credit;
+    obj.middleDebit = 0;
+    obj.middleCredit = 0;
+    obj.is_charge = row.is_charge;
+    obj.is_asset = row.is_asset;
+    account[id] = obj;
+    return account;
+  }, {});
+
+  // format and process the monthly balance for accounts
+  balances.middle.forEach(function (row){
+    var account = accounts[row.number] || {};
+    account.middleDebit = row.debit;
+    account.middleCredit = row.credit;
+    account.label = row.label;
+    account.is_charge = row.is_charge;
+    account.is_asset = row.is_asset;
+    accounts[row.number] = account;
+  });
+
+  Object.keys(accounts).forEach(function (item){
+    accounts[item].endDebit = 0;
+    accounts[item].endCredit = 0;
+    var sold = (accounts[item].beginDebit || 0 - accounts[item].beginCredit || 0) + (accounts[item].middleDebit - accounts[item].middleCredit);
+    if(sold < 0){
+      accounts[item].endCredit = sold * -1;
+    }else{
+     accounts[item].endDebit = sold;
+    }
+  });
+
+  // process for getting totals
+  totals = Object.keys(accounts).reduce(function (totals, key) {
+    var account = accounts[key];
+    totals.beginDebit += (account.beginDebit || 0);
+    totals.beginCredit += (account.beginCredit || 0);
+    totals.middleDebit += (account.middleDebit || 0);
+    totals.middleCredit += (account.middleCredit || 0);
+    totals.endDebit += (account.endDebit || 0);
+    totals.endCredit += (account.endCredit || 0);
+    return totals;
+  }, {
+    beginDebit : 0,
+    beginCredit : 0,
+    middleDebit: 0,
+    middleCredit : 0,
+    endDebit : 0,
+    endCredit : 0
+  });
+
+  return { accounts, totals };
+}
+
+/**
+ * @function getSold
+ * @description return the balance of an account
+ * @param {object} object An object from the array returned by balanceReporting function
+ */
+function getSold (item){
+  let debit  = 0;
+  let credit = 0;
+  let sold   = 0;
+
+  if(item.is_asset === 1 || item.is_charge === 1){
+    sold = item.debit - item.credit;
+    if(sold < 0){
+      credit = sold * -1 ;
+    }else{
+      debit = sold;
+    }
+  }else{
+    sold = item.credit - item.debit;
+    if(sold < 0){
+      debit = sold * -1;
+    } else{
+      credit = sold;
+    }
+  }
+  return { debit, credit };
+}
+
+/**
+ * @function balanceReporting
+ * @description return the balance needed according the given params
+ * @param {object} params An object which contains dates range and the account class
+ */
+function balanceReporting(params) {
+  let sql, hasClasse, dateRange, queryParameters,
+      query = params,
+      data = {};
+
+  hasClasse = (query.classe !== '*');
+  dateRange = (query.dateFrom && query.dateTo);
+
+  // gets the amount up to the current period
+  sql =
+    'SELECT a.number, a.id, a.label, a.type_id, a.is_charge, a.is_asset, SUM(pt.credit) AS credit, SUM(pt.debit) AS debit ' +
```

Comment:
These would be easier to read and modify if you used template strings in the future.

```js
sql = `
  SELECT a.number, a.id, a.label, a.type_id, a.is_charge, a.is_asset, SUM(pt.credit) AS credit, SUM(pt.debit) AS debit
  FROM period_total AS pt JOIN account AS a ON pt.account_id = a.id
  JOIN period AS p ON pt.period_id = p.id 
  WHERE p.end_date <= DATE(?) AND pt.enterprise_id = ?
  ${hasClasse ? 'AND a.classe = ? ' : '' }
  GROUP BY a.id;
`;

That way, we can copy this string in to the mysql command prompt when we need to debug.